dnf5/commands/check/check.cpp

```diff
@@ -0,0 +1,311 @@
+    auto * arch = headerGetString(h, RPMTAG_ARCH);
+    if (!arch) {
+        // skip special package representing key installed in the rpm database
+        return 0;
+    }
+
+    auto * name = headerGetString(h, RPMTAG_NAME);
+
+    auto & nevras = installed_na_nevras[fmt::format("{}.{}", name, arch)];
+    if (nevras.empty() || !is_package_installonly(h)) {
+        // "duplicate" packages are only added if they are not "installonly" packages
+        // to avoid being reported as duplicates.
+        std::unique_ptr<char, decltype(&rfree)> nevra(headerGetAsString(h, RPMTAG_NEVRA), &rfree);
+        nevras.push_back(nevra.get());
+    }
+
+    return 0;
+}
+
+}  // namespace
+
+
+void CheckCommand::set_parent_command() {
+    auto * arg_parser_parent_cmd = get_session().get_argument_parser().get_root_command();
+    auto * arg_parser_this_cmd = get_argument_parser_command();
+    arg_parser_parent_cmd->register_command(arg_parser_this_cmd);
+    arg_parser_parent_cmd->get_group("software_management_commands").register_argument(arg_parser_this_cmd);
+}
+
+
+void CheckCommand::set_argument_parser() {
+    auto & ctx = get_context();
+    auto & parser = ctx.get_argument_parser();
+
+    auto & cmd = *get_argument_parser_command();
+    cmd.set_description("Check for problems in the packagedb");
+
+    cmd.register_named_arg(create_option(parser, "dependencies", "Show dependency problems", dependencies));
+    cmd.register_named_arg(create_option(parser, "duplicates", "Show duplicate problems", duplicates));
+    cmd.register_named_arg(create_option(parser, "obsoleted", "Show obsoleted package", obsoleted));
+}
+
+
+void CheckCommand::configure() {
+    installonly_pkgs_provides = get_context().base.get_config().get_installonlypkgs_option().get_value();
+    all = !(dependencies || duplicates || obsoleted);
+}
+
+
+void CheckCommand::run() {
+    auto & ctx = get_context();
+
+    auto ts = rpmtsCreate();
+    auto & config = ctx.base.get_config();
+    rpmtsSetRootDir(ts, config.get_installroot_option().get_value().c_str());
+
+    if (all || dependencies || obsoleted) {
+        QVA_t qva = &rpmQVKArgs;
+        qva->qva_mode = 'V';
+        qva->qva_flags = VERIFY_DEPS;
+        qva->qva_source = RPMQV_ALL;
+        qva->qva_sourceCount = 1;
+        qva->qva_showPackage = find_problems;
+
+        struct rpmInstallArguments_s * ia = &rpmIArgs;
+        rpmtsSetFlags(ts, rpmtsFlags(ts) | (ia->transFlags & RPMTRANS_FLAG_NOPLUGINS));
+
+        rpmcliVerify(ts, qva, NULL);
+    }
+
+    if (all || duplicates) {
+        QVA_t qva = &rpmQVKArgs;
+        qva->qva_mode = 'q';
+        qva->qva_flags = VERIFY_DEPS;
+        qva->qva_source = RPMQV_ALL;
+        qva->qva_sourceCount = 1;
+        qva->qva_showPackage = group_pkgs_same_name_arch;
+
+        struct rpmInstallArguments_s * ia = &rpmIArgs;
+        rpmtsSetFlags(ts, rpmtsFlags(ts) | (ia->transFlags & RPMTRANS_FLAG_NOPLUGINS));
+
+        rpmcliQuery(ts, qva, NULL);
```

Comment:
I'm a bit surprised to see rpmcliVerify() and rpmcliQuery() used, these cli-variants of the API aren't particularly popular amongst developers...

Not that it's wrong, but if it was me, I would've probably just used something like

```
Header h = NULL;
rpmdbMatchIterator mi = rpmtsInitIterator(ts, RPMDBI_PACKAGES, NULL, 0);
while ((h = rpmdbNextIterator(mi))) {
    find_problems(ts, h);
}
rpmdbFreeIterator(mi);
```

...and similarly for group_pkgs_same_name_arch.

Reply:
@pmatilai To be honest, I don't like using `rpmcliVerify` and `rpmcliQuery`. But those functions do more than just iterate.
Example from `rpmciVerify`:
```
    rpmtsOpenDB(ts, O_RDONLY);
    rpmdbOpenAll(rpmtsGetRdb(ts));
    if (rpmChrootSet(rpmtsRootDir(ts)) || rpmChrootIn()) {
```
and also calls `rpmtsSetVSFlags`.

I don't know why all this is being done. That's why I've used higher functions so far. And in the future, when I understand it, I will rewrite this piece of code. I assume when adding file checking support. But if it's as easy as you suggest, I can do it now.

Reply:
Those functions do all manner of "generic" stuff that's probably not relevant to you. But hmm, rpmtsCheck() can end up looking into the filesystem so if you need "check" to work with --installroot then (at least some of) that chroot dance done by rpmcliVerify() is necessary :thinking: